Classes/common/query/CDTQIndex.m

```diff
@@ -0,0 +1,171 @@
+//
+//  CDTQIndex.m
+//
+//  Created by Al Finkelstein on 2015-04-20
+//  Copyright (c) 2015 Cloudant. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+//  except in compliance with the License. You may obtain a copy of the License at
+//    http://www.apache.org/licenses/LICENSE-2.0
+//  Unless required by applicable law or agreed to in writing, software distributed under the
+//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+//  either express or implied. See the License for the specific language governing permissions
+//  and limitations under the License.
+
+#import "CDTQIndex.h"
+
+#import "CDTQLogging.h"
+
+NSString *const kCDTQJsonType = @"json";
+NSString *const kCDTQTextType = @"text";
+
+static NSString *const kCDTQTextTokenize = @"tokenize";
+static NSString *const kCDTQTextDefaultTokenizer = @"simple";
+
+@interface CDTQIndex ()
+
+@end
+
+@implementation CDTQIndex
+
+// Static array of supported index types
++ (NSArray *)validTypes
+{
+    static NSArray *validTypesArray = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        validTypesArray = @[ kCDTQJsonType, kCDTQTextType ];
+    });
+    return validTypesArray;
+}
+
+// Static array of supported index settings
++ (NSArray *)validSettings
+{
+    static NSArray *validSettingsArray = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        validSettingsArray = @[ kCDTQTextTokenize ];
+    });
+    return validSettingsArray;
+}
+
+- (instancetype)initWithFields:(NSArray *)fieldNames
+                     indexName:(NSString *)indexName
+                     indexType:(NSString *)indexType
+                 indexSettings:(NSDictionary *)indexSettings
+{
+    self = [super init];
+    if (self) {
+        _fieldNames = fieldNames;
+        _indexName = indexName;
+        _indexType = indexType;
+        _indexSettings = indexSettings;
+    }
+    return self;
+}
+
++ (CDTQIndex *)index:(NSString *)indexName withFields:(NSArray *)fieldNames
+{
+    return [CDTQIndex index:indexName withFields:fieldNames ofType:kCDTQJsonType];
+}
```

Comment:
For these three convenience constructors, we should return `instancetype`. In addition, we should call the constructor with more arguments using `[[self class] index:...]. Finally, for the `-initWithFields:...` method at the end of the final convenience constructor, use `[[self class] alloc]...`.

This allows sub-classes to override these methods and have the override method be called.

See the instancetype discussion here: https://developer.apple.com/library/ios//releasenotes/ObjectiveC/ModernizationObjC/AdoptingModernObjective-C/AdoptingModernObjective-C.html.